### PR TITLE
<v2.9> Convert test_wl_upgrade to go

### PR DESCRIPTION
**Pull request by lscalabrini01**
Implemented TestWorkloadUpgrade test

### CODE REVIEW

**Comment by slickwarren** [comment]
some conflicts are popping up but otherwise lgtm

File: tests/v2/actions/workloads/deployment/deployment.go
Comment (slickwarren) [inline tests/v2/actions/workloads/deployment/deployment.go:102]: all you need is the deployment name, right? not the whole deployment?

File: tests/v2/validation/workloads/workload_test.go
Comment (slickwarren) [inline tests/v2/validation/workloads/workload_test.go:240]: imo deepcopying the deployment is a little confusing. Because you're updating the same rancher object each time, but storing different objects for nginx and redis in the code. imo it would be better to update the same object each time instead of copying it.

File: tests/v2/validation/workloads/workload_test.go
Comment (slickwarren) [inline tests/v2/validation/workloads/workload_test.go:274]: same comment here

File: tests/v2/validation/workloads/workload_test.go
Comment (slickwarren) [inline tests/v2/validation/workloads/workload_test.go:293]: this is repeated quite a bit in here. Would it be worth writing a short helper function since these 2 are called in tandem so often? something like 'ValidateRolloutHistory(client, expectedRevisions, expectedContainerCount) error {...}`

File: tests/v2/validation/workloads/workload_test.go
Comment (slickwarren) [inline tests/v2/validation/workloads/workload_test.go:317]: just pointing out that, the deployment isn't used in rollbackDeployment, just its name. so this is sort of confusing just reading through the code when using copied versions of the deployment vs. just using the same deployment / deploymentName and requesting to rollback to specific revisions

File: tests/v2/actions/workloads/deployment/deployment.go
Comment (slickwarren) [inline tests/v2/actions/workloads/deployment/deployment.go:111]: consolidating my other comments; maybe this could be validateDeploymentRolloutHistory(client, clusterID, namespace, deploymentName, expectedRolloutHistory) error {
...
}
since we usually don't have `get` function helpers as actions

**Review by slickwarren — COMMENTED** [review]
sorry, I had pending comments in this PR instead of main, so I accidentally approved there. Since that has merged feel free to address these comments in a separate PR.

File: tests/v2/actions/workloads/deployment/deployment.go
Comment (lscalabrini01) [reply tests/v2/actions/workloads/deployment/deployment.go:102]: Yes, I can change the parameters to get the deployment Name, but I also like the idea of pass the deployment and function is responsible to get the needed deployment informations

File: tests/v2/validation/workloads/workload_test.go
Comment (lscalabrini01) [reply tests/v2/validation/workloads/workload_test.go:240]: I agree

File: tests/v2/validation/workloads/workload_test.go
Comment (lscalabrini01) [reply tests/v2/validation/workloads/workload_test.go:293]: It's a good ideia, I will create a private function in the workloads_test

**Review by floatingman — DISMISSED** [review]
LGTM

File: go.mod
Comment (slickwarren) [inline go.mod:144]: is this expected?

File: tests/v2/actions/workloads/deployment/deployment.go
Comment (slickwarren) [reply tests/v2/actions/workloads/deployment/deployment.go:102]: okay I'll leave it up to you then. But usually we want to keep the params within scope, so if it isn't using the full deployment it would make sense to just send what the func needs.

**Review by slickwarren — COMMENTED** [review]
lgtm, just some follow up questions

File: tests/v2/validation/workloads/workload_test.go
Comment (anupama2501) [inline tests/v2/validation/workloads/workload_test.go:278]: since this is not rollout history validation, can we move this to the test itself or change the function name to something else? its a bit misleading with validateRolloutHistory

File: tests/v2/actions/workloads/deployment/deployment.go
Comment (anupama2501) [inline tests/v2/actions/workloads/deployment/deployment.go:114]: This just caught my attention. Instead of using the command, why don’t we retrieve the deployment and get the revision history either using Steve v1 or the public API? Using the command is fine, but it’s a workaround to execute it via a job. Directly using the API would be more convenient.

File: tests/v2/validation/workloads/workload_test.go
Comment (anupama2501) [inline tests/v2/validation/workloads/workload_test.go:260]: once we change from command to v1 or public api to get the deployment and then revision history, we wouldn't need to send in the values as a list of strings here.

File: tests/v2/actions/workloads/deployment/deployment.go
Comment (lscalabrini01) [reply tests/v2/actions/workloads/deployment/deployment.go:102]: I'll add the parameter

File: tests/v2/validation/workloads/workload_test.go
Comment (lscalabrini01) [reply tests/v2/validation/workloads/workload_test.go:278]: Yes, I can

File: tests/v2/actions/workloads/deployment/deployment.go
Comment (lscalabrini01) [reply tests/v2/actions/workloads/deployment/deployment.go:114]: Because the python test is using the command line to get rollout history and I also need to get the all revisions and using the API it would get just the last revison by the `RevisionHistoryLimit` field

File: tests/v2/validation/workloads/workload_test.go
Comment (lscalabrini01) [reply tests/v2/validation/workloads/workload_test.go:260]: Yes, I can do this change, but it would not be the same test of test_wl_upgrade

File: tests/v2/actions/workloads/deployment/deployment.go
Comment (lscalabrini01) [reply tests/v2/actions/workloads/deployment/deployment.go:114]: the RevisionHistoryLimit is always returning 10 and I didn't find any other field that I could use to get the revision, may be for while we could maintain the command

**Review by floatingman — DISMISSED** [review]
LGTM

File: tests/v2/actions/workloads/deployment/deployment.go
Comment (slickwarren) [reply tests/v2/actions/workloads/deployment/deployment.go]: last point here but won't block on it; if you only ever `GetRolloutHistoryDeployment` then immediately `validateRolloutHistory` then  I would vote to just loop it all into one function since we usually don't add actions that only `Get`. does that make sense?

**Review by slickwarren — DISMISSED** [review]

**Review by slickwarren — DISMISSED** [review]

**Review by floatingman — APPROVED** [review]
LGTM

**Review by slickwarren — APPROVED** [review]

**Pull request merged**
